File: frontend/src/services/tasks.ts
// frontend/src/services/tasks.ts
import { getToken, isAuthenticated } from '@/lib/auth';

interface Task {
  id: number; // Backend uses numeric IDs
  title: string;
  description?: string;
  completed: boolean;
  user_id: string; // Backend uses snake_case
  created_at: string; // Backend uses snake_case
  updated_at: string; // Backend uses snake_case
}

interface TaskCreateRequest {
  title: string;
  description?: string;
}

interface TaskUpdateRequest {
  title?: string;
  description?: string;
  completed?: boolean;
}

// Base API URL - this should come from environment variables
const API_BASE_URL = process.env.NEXT_PUBLIC_API_BASE_URL || 'http://127.0.0.1:8000';

/**
 * Helper function to make authenticated API requests
 */
const apiRequest = async (endpoint: string, options: RequestInit = {}) => {
  // Check if user is authenticated
  const authenticated = isAuthenticated();
  if (!authenticated) {
    console.error('User is not authenticated');
    throw new Error('User is not authenticated');
  }

  // Get the token
  const token = getToken();
  if (!token) {
    console.error('No token found');
    throw new Error('No authentication token found');
  }

  const url = `${API_BASE_URL}${endpoint}`;

  const headers = {
    'Content-Type': 'application/json',
    'Authorization': `Bearer ${token}`,
    ...options.headers,
  };

  // Log the request for debugging (remove in production)
  console.log('Making API request:', {
    url,
    headers: {
      'Content-Type': headers['Content-Type'],
      'Authorization': headers['Authorization'] ? 'Bearer [TOKEN]' : 'Missing'
    }
  });

  const response = await fetch(url, {
    ...options,
    headers,
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    console.error('API request failed:', errorData);
    throw new Error(errorData.detail || `HTTP error! status: ${response.status}`);
  }

  // For DELETE requests, there's typically no response body
  if (response.status === 204) {
    return null;
  }

  return response.json();
};

/**
 * Get all tasks for a user
 * @param userId - The ID of the user whose tasks to retrieve
 * @returns Promise resolving to array of tasks
 */
export const getTasks = async (userId: string): Promise<Task[]> => {
  try {
    // Validate userId to prevent double slash in URL
    if (!userId) {
      throw new Error('User ID is required to fetch tasks');
    }
    const tasks = await apiRequest(`/api/${userId}/tasks`);
    return tasks;
  } catch (error) {
    console.error('Error fetching tasks:', error);
    throw error;
  }
};

/**
 * Create a new task
 * @param userId - The ID of the user creating the task
 * @param taskData - The task data to create
 * @returns Promise resolving to the created task
 */
export const createTask = async (userId: string, taskData: TaskCreateRequest): Promise<Task> => {
  try {
    // Validate userId to prevent double slash in URL
    if (!userId) {
      throw new Error('User ID is required to create a task');
    }
    const task = await apiRequest(`/api/${userId}/tasks`, {
      method: 'POST',
      body: JSON.stringify(taskData),
    });
    return task;
  } catch (error) {
    console.error('Error creating task:', error);
    throw error;
  }
};

/**
 * Update an existing task
 * @param userId - The ID of the user who owns the task
 * @param taskId - The ID of the task to update
 * @param taskData - The updated task data
 * @returns Promise resolving to the updated task
 */
export const updateTask = async (userId: string, taskId: string, taskData: TaskUpdateRequest): Promise<Task> => {
  try {
    // Validate userId to prevent double slash in URL
    if (!userId) {
      throw new Error('User ID is required to update a task');
    }
    const task = await apiRequest(`/api/${userId}/tasks/${taskId}`, {
      method: 'PUT',
      body: JSON.stringify(taskData),
    });
    return task;
  } catch (error) {
    console.error('Error updating task:', error);
    throw error;
  }
};

/**
 * Delete a task
 * @param userId - The ID of the user who owns the task
 * @param taskId - The ID of the task to delete
 * @returns Promise resolving when task is deleted
 */
export const deleteTask = async (userId: string, taskId: string): Promise<void> => {
  try {
    // Validate userId to prevent double slash in URL
    if (!userId) {
      throw new Error('User ID is required to delete a task');
    }
    await apiRequest(`/api/${userId}/tasks/${taskId}`, {
      method: 'DELETE',
    });
  } catch (error) {
    console.error('Error deleting task:', error);
    throw error;
  }
};

/**
 * Toggle task completion status
 * @param userId - The ID of the user who owns the task
 * @param taskId - The ID of the task to update completion status
 * @param completed - The new completion status
 * @returns Promise resolving to the updated task
 */
export const toggleComplete = async (userId: string, taskId: string, completed: boolean): Promise<Task> => {
  try {
    // Validate userId to prevent double slash in URL
    if (!userId) {
      throw new Error('User ID is required to toggle task completion');
    }
    const task = await apiRequest(`/api/${userId}/tasks/${taskId}/complete`, {
      method: 'PATCH',
    });
    return task;
  } catch (error) {
    console.error('Error toggling task completion:', error);
    throw error;
  }
};

/**
 * Get the count of pending tasks for a user
 * @param userId - The ID of the user whose pending tasks to count
 * @returns Promise resolving to an object with the pending count
 */
export interface PendingTasksCount {
  pending: number;
}

export const getPendingTasksCount = async (userId: string): Promise<PendingTasksCount> => {
  try {
    // Validate userId to prevent double slash in URL
    if (!userId) {
      throw new Error('User ID is required to fetch pending tasks count');
    }
    const result = await apiRequest(`/api/${userId}/pending-tasks`);
    return result;
  } catch (error) {
    console.error('Error fetching pending tasks count:', error);
    throw error;
  }
};

/**
 * Get the count of completed tasks for a user
 * @param userId - The ID of the user whose completed tasks to count
 * @returns Promise resolving to an object with the completed count
 */
export interface CompletedTasksCount {
  completed: number;
}

export const getCompletedTasksCount = async (userId: string): Promise<CompletedTasksCount> => {
  try {
    // Validate userId to prevent double slash in URL
    if (!userId) {
      throw new Error('User ID is required to fetch completed tasks count');
    }
    const result = await apiRequest(`/api/${userId}/completed-tasks`);
    return result;
  } catch (error) {
    console.error('Error fetching completed tasks count:', error);
    throw error;
  }
};

File: backend/src/api/routes/tasks.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlmodel import Session
from typing import List
import logging
from src.database.session import get_session
from src.api.deps import verify_user_access
from src.models.task import Task, TaskCreate, TaskRead, TaskUpdate
from src.services.task_service import TaskService

logger = logging.getLogger(__name__)

router = APIRouter()

@router.post("/tasks", response_model=TaskRead, status_code=status.HTTP_201_CREATED)
async def create_task(
    task_data: TaskCreate,
    payload: dict = Depends(verify_user_access),
    session: Session = Depends(get_session)
):
    """
    Create a new task for the authenticated user.
    """
    # Get user_id from the verified access function
    user_id = payload.get("userId") or payload.get("sub")

    # Create the task using the service
    task = TaskService.create_task(
        session=session,
        user_id=user_id,
        title=task_data.title,
        description=task_data.description
    )

    return task


@router.get("/tasks", response_model=List[TaskRead])
async def get_user_tasks(
    payload: dict = Depends(verify_user_access),
    session: Session = Depends(get_session)
):
    """
    Get all tasks for the authenticated user.
    """
    # Get user_id from the verified access function
    user_id = payload.get("userId") or payload.get("sub")

    # Get tasks for the user
    tasks = TaskService.get_user_tasks(session=session, user_id=user_id)

    return tasks


@router.get("/tasks/{id}", response_model=TaskRead)
async def get_task(
    id: int,
    payload: dict = Depends(verify_user_access),
    session: Session = Depends(get_session)
):
    """
    Get a specific task by ID for the authenticated user.
    """
    # Get user_id from the verified access function
    user_id = payload.get("userId") or payload.get("sub")

    # Get the specific task
    task = TaskService.get_task_by_id(session=session, task_id=id, user_id=user_id)

    if not task:
        raise HTTPException(status_code=404, detail="Task not found")

    return task


@router.put("/tasks/{id}", response_model=TaskRead)
async def update_task(
    id: int,
    task_update: TaskUpdate,
    payload: dict = Depends(verify_user_access),
    session: Session = Depends(get_session)
):
    """
    Update a specific task for the authenticated user.
    """
    # Get user_id from the verified access function
    user_id = payload.get("userId") or payload.get("sub")

    logger.info(f"PUT /tasks/{id} - Request for user {user_id}")

    # Update the task
    updated_task = TaskService.update_task(
        session=session,
        task_id=id,
        user_id=user_id,
        task_update=task_update
    )

    if not updated_task:
        logger.warning(f"PUT /tasks/{id} - Task not found for user {user_id}")
        raise HTTPException(status_code=404, detail="Task not found")

    logger.info(f"PUT /tasks/{id} - Task updated successfully for user {user_id}")
    return updated_task


@router.delete("/tasks/{id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_task(
    id: int,
    payload: dict = Depends(verify_user_access),
    session: Session = Depends(get_session)
):
    """
    Delete a specific task for the authenticated user.
    """
    # Get user_id from the verified access function
    user_id = payload.get("userId") or payload.get("sub")

    logger.info(f"DELETE /tasks/{id} - Request for user {user_id}")

    # Delete the task
    deleted = TaskService.delete_task(session=session, task_id=id, user_id=user_id)

    if not deleted:
        logger.warning(f"DELETE /tasks/{id} - Task not found for user {user_id}")
        raise HTTPException(status_code=404, detail="Task not found")

    logger.info(f"DELETE /tasks/{id} - Task deleted successfully for user {user_id}")
    return {"message": "Task deleted successfully"}


@router.patch("/tasks/{id}/complete", response_model=TaskRead)
async def toggle_task_completion(
    id: int,
    payload: dict = Depends(verify_user_access),
    session: Session = Depends(get_session)
):
    """
    Toggle the completion status of a task for the authenticated user.
    """
    # Get user_id from the verified access function
    user_id = payload.get("userId") or payload.get("sub")

    logger.info(f"PATCH /tasks/{id}/complete - Request for user {user_id}")

    # Toggle task completion
    task = TaskService.toggle_completion(session=session, task_id=id, user_id=user_id)

    if not task:
        logger.warning(f"PATCH /tasks/{id}/complete - Task not found for user {user_id}")
        raise HTTPException(status_code=404, detail="Task not found")

    logger.info(f"PATCH /tasks/{id}/complete - Task completion toggled successfully for user {user_id}")
    return task

@router.get("/pending-tasks", response_model=dict)
async def get_pending_tasks_count(
    payload: dict = Depends(verify_user_access),
    session: Session = Depends(get_session)
):
    """
    Get the count of pending tasks for the authenticated user.
    Pending tasks are tasks that are not completed.
    """
    # Get user_id from the verified access function
    user_id = payload.get("userId") or payload.get("sub")

    logger.info(f"GET /pending-tasks - Request for user {user_id}")

    # Get the count of pending tasks
    pending_count = TaskService.get_pending_tasks_count(session=session, user_id=user_id)

    logger.info(f"GET /pending-tasks - Returning count: {pending_count} for user {user_id}")

    return {"pending": pending_count}

@router.get("/completed-tasks", response_model=dict)
async def get_completed_tasks_count(
    payload: dict = Depends(verify_user_access),
    session: Session = Depends(get_session)
):
    """
    Get the count of completed tasks for the authenticated user.
    Completed tasks are tasks where completed = True.
    """
    # Get user_id from the verified access function
    user_id = payload.get("userId") or payload.get("sub")

    logger.info(f"GET /completed-tasks - Request for user {user_id}")

    # Get the count of completed tasks
    completed_count = TaskService.get_completed_tasks_count(session=session, user_id=user_id)

    logger.info(f"GET /completed-tasks - Returning count: {completed_count} for user {user_id}")

    return {"completed": completed_count}